File: CS/WindowsFormsApplication1/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraScheduler;

namespace WindowsFormsApplication1
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        TimeSpan minTime = new TimeSpan(8, 0, 0);
        TimeSpan maxTime = new TimeSpan(19, 0, 0);

        private void Form1_Load(object sender, EventArgs e)
        {
            TimeScaleCollection scales = schedulerControl1.TimelineView.Scales;
            schedulerControl1.TimelineView.Scales.BeginUpdate();
            try
            {
                scales.Clear();
                scales.Add(new MyTimeScaleFixedInterval(TimeSpan.FromDays(1), minTime, maxTime));
                scales.Add(new MyTimeScaleFixedInterval(TimeSpan.FromHours(1), minTime, maxTime));
                scales.Add(new MyTimeScaleFixedInterval(TimeSpan.FromMinutes(30), minTime, maxTime));
            }
            finally
            {
                schedulerControl1.TimelineView.Scales.EndUpdate();
            }

            Appointment apt = schedulerStorage1.CreateAppointment(AppointmentType.Normal);
            apt.Start = DateTime.Today.AddHours(10);
            apt.End = DateTime.Today.AddHours(14);
            apt.Subject = "Test appointment";
            schedulerStorage1.Appointments.Add(apt);
        }

        void CorrectAppointmentParameters(AppointmentDragEventArgs e)
        {
            TimeSpan customScaleDuration = (maxTime - minTime).Add(TimeSpan.FromHours(2));
            DateTime initialStart = e.EditedAppointment.Start;

            if (e.EditedAppointment.Start.TimeOfDay < minTime || e.EditedAppointment.Start.TimeOfDay >= maxTime)
            {
                bool durationMustBeIncreased = e.EditedAppointment.Duration > customScaleDuration;
                e.EditedAppointment.Start = durationMustBeIncreased ? e.EditedAppointment.Start + customScaleDuration : e.EditedAppointment.Start - customScaleDuration;

                TimeInterval appInterval = new TimeInterval(e.EditedAppointment.Start, e.EditedAppointment.Start + e.EditedAppointment.Duration + customScaleDuration);
                if (!appInterval.Contains(e.HitInterval))
                {
                    durationMustBeIncreased = !durationMustBeIncreased;
                    e.EditedAppointment.Start = durationMustBeIncreased ? initialStart + customScaleDuration : initialStart - customScaleDuration;
                }
                e.Handled = true;
            }

            if (e.EditedAppointment.End.TimeOfDay >= maxTime || e.EditedAppointment.End.TimeOfDay < minTime)
            {
                e.EditedAppointment.End = e.EditedAppointment.Duration > customScaleDuration ? e.EditedAppointment.End - customScaleDuration : e.EditedAppointment.End + customScaleDuration;
                e.Handled = true;
            }         
        }

        private void schedulerControl1_AppointmentDrag(object sender, AppointmentDragEventArgs e)
        {
            CorrectAppointmentParameters(e);
        }

        private void schedulerControl1_AppointmentDrop(object sender, AppointmentDragEventArgs e)
        {
            CorrectAppointmentParameters(e);
        }
    }
}

File: VB/WindowsFormsApplication1/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.XtraScheduler

Namespace WindowsFormsApplication1
    Partial Public Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private minTime As New TimeSpan(8, 0, 0)
        Private maxTime As New TimeSpan(19, 0, 0)

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            Dim scales As TimeScaleCollection = schedulerControl1.TimelineView.Scales
            schedulerControl1.TimelineView.Scales.BeginUpdate()
            Try
                scales.Clear()
                scales.Add(New MyTimeScaleFixedInterval(TimeSpan.FromDays(1), minTime, maxTime))
                scales.Add(New MyTimeScaleFixedInterval(TimeSpan.FromHours(1), minTime, maxTime))
                scales.Add(New MyTimeScaleFixedInterval(TimeSpan.FromMinutes(30), minTime, maxTime))
            Finally
                schedulerControl1.TimelineView.Scales.EndUpdate()
            End Try

            Dim apt As Appointment = schedulerStorage1.CreateAppointment(AppointmentType.Normal)
            apt.Start = Date.Today.AddHours(10)
            apt.End = Date.Today.AddHours(14)
            apt.Subject = "Test appointment"
            schedulerStorage1.Appointments.Add(apt)
        End Sub

        Private Sub CorrectAppointmentParameters(ByVal e As AppointmentDragEventArgs)
            Dim customScaleDuration As TimeSpan = (maxTime.Subtract(minTime)).Add(TimeSpan.FromHours(2))
            Dim initialStart As Date = e.EditedAppointment.Start

            If e.EditedAppointment.Start.TimeOfDay < minTime OrElse e.EditedAppointment.Start.TimeOfDay >= maxTime Then
                Dim durationMustBeIncreased As Boolean = e.EditedAppointment.Duration > customScaleDuration
                e.EditedAppointment.Start = If(durationMustBeIncreased, e.EditedAppointment.Start + customScaleDuration, e.EditedAppointment.Start - customScaleDuration)

                Dim appInterval As New TimeInterval(e.EditedAppointment.Start, e.EditedAppointment.Start + e.EditedAppointment.Duration + customScaleDuration)
                If Not appInterval.Contains(e.HitInterval) Then
                    durationMustBeIncreased = Not durationMustBeIncreased
                    e.EditedAppointment.Start = If(durationMustBeIncreased, initialStart.Add(customScaleDuration), initialStart.Subtract(customScaleDuration))
                End If
                e.Handled = True
            End If

            If e.EditedAppointment.End.TimeOfDay >= maxTime OrElse e.EditedAppointment.End.TimeOfDay < minTime Then
                e.EditedAppointment.End = If(e.EditedAppointment.Duration > customScaleDuration, e.EditedAppointment.End - customScaleDuration, e.EditedAppointment.End + customScaleDuration)
                e.Handled = True
            End If
        End Sub

        Private Sub schedulerControl1_AppointmentDrag(ByVal sender As Object, ByVal e As AppointmentDragEventArgs) Handles schedulerControl1.AppointmentDrag
            CorrectAppointmentParameters(e)
        End Sub

        Private Sub schedulerControl1_AppointmentDrop(ByVal sender As Object, ByVal e As AppointmentDragEventArgs) Handles schedulerControl1.AppointmentDrop
            CorrectAppointmentParameters(e)
        End Sub
    End Class
End Namespace
